lightning/src/routing/onion_message.rs

```diff
@@ -0,0 +1,283 @@
+}
+
+#[derive(Debug, PartialEq)]
+/// Errors that might occur running [`find_path`].
+pub enum Error {
+	/// We failed to find a path to the destination.
+	PathNotFound,
+	/// We failed to convert this node id into a [`PublicKey`].
+	InvalidNodeId(secp256k1::Error),
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		match self {
+			Error::PathNotFound => write!(f, "Failed to find a path to the destination"),
+			Error::InvalidNodeId(e) =>
+				write!(f, "Failed to convert a node id into a PublicKey with error: {}", e),
+		}
+	}
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+#[derive(Eq, PartialEq)]
+struct PathBuildingHop {
+	cost: u64,
+	node_id: NodeId,
+	parent_node_id: NodeId,
+}
+
+impl PartialOrd for PathBuildingHop {
+	fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+		// We need a min-heap, whereas `BinaryHeap`s are a max-heap, so compare the costs in reverse.
+		other.cost.partial_cmp(&self.cost)
+	}
+}
+
+impl Ord for PathBuildingHop {
+	fn cmp(&self, other: &Self) -> cmp::Ordering {
+		self.partial_cmp(other).unwrap()
+	}
+}
+
+fn reverse_path(
+	parents: HashMap<NodeId, NodeId>, our_node_id: NodeId, destination: NodeId
+)-> Result<Vec<PublicKey>, Error>
+{
+	let mut path = Vec::new();
+	let mut curr = destination;
+	loop {
+		match PublicKey::from_slice(curr.as_slice()) {
+			Ok(pk) => path.push(pk),
+			Err(e) => return Err(Error::InvalidNodeId(e))
+		}
+		match parents.get(&curr) {
+			None => return Err(Error::PathNotFound),
+			Some(parent) => {
+				if *parent == our_node_id { break; }
+				curr = *parent;
+			}
+		}
+	}
+
+	path.reverse();
+	Ok(path)
+}
+
+#[cfg(test)]
+mod tests {
+	use ln::features::{InitFeatures, NodeFeatures};
+	use routing::test_utils::{add_or_update_node, build_graph_with_features, build_line_graph, get_channel_details, get_nodes};
+
+	use sync::Arc;
+
+	#[test]
+	fn one_hop() {
+		let mut features = NodeFeatures::empty();
+		features.set_onion_messages_optional();
+		let (secp_ctx, network_graph, _, _, logger) = build_graph_with_features(features);
+		let (_, our_id, _, node_pks) = get_nodes(&secp_ctx);
+
+		let path = super::find_path(&our_id, &node_pks[7], &network_graph, None, Arc::clone(&logger)).unwrap();
+		assert_eq!(path.len(), 1);
+		assert!(path[0] == node_pks[7]);
+	}
+
+	#[test]
+	fn two_hops() {
+		let mut features = NodeFeatures::empty();
+		features.set_onion_messages_optional();
+		let (secp_ctx, network_graph, _, _, logger) = build_graph_with_features(features);
+		let (_, our_id, _, node_pks) = get_nodes(&secp_ctx);
+
+		let path = super::find_path(&our_id, &node_pks[2], &network_graph, None, Arc::clone(&logger)).unwrap();
+		assert_eq!(path.len(), 2);
+		// See routing::test_utils::build_graph ASCII graph, the first hop can be any of these
+		assert!(path[0] == node_pks[1] || path[0] == node_pks[7] || path[0] == node_pks[0]);
+		assert_eq!(path[1], node_pks[2]);
+	}
+
+	#[test]
+	fn three_hops() {
```

Comment:
Not sure whether this test case is not a superset of `one_hop`/`two_hops`? Do we need all three, or can we just keep this one?